File: java/rest_assured/src/test/java/rest_api/CatalogTest.java
package rest_api;

import io.qameta.allure.Allure;
import io.qameta.allure.restassured.AllureRestAssured;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;

import io.restassured.RestAssured;
import io.restassured.builder.RequestSpecBuilder;
import io.restassured.http.ContentType;
import io.restassured.path.json.JsonPath;
import io.restassured.response.Response;
import io.restassured.specification.RequestSpecification;
import static io.restassured.RestAssured.*;
import static org.hamcrest.Matchers.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;


@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
public class CatalogTest {

    static String id = null;


    @BeforeAll
    public static void setUp() {
        RestAssured.baseURI = "https://api.restful-api.dev";
        RestAssured.basePath = "/objects";
        RestAssured.requestSpecification = 
            (RequestSpecification) new RequestSpecBuilder()
                .setContentType(ContentType.JSON)
                .setAccept(ContentType.JSON)
                .build();  //.log().all();  //log().uri().log().method();
    }

    @Test
    @Order(1)
    public void add_object() {
        Allure.getLifecycle().updateTestCase(tr -> tr.getLabels().removeIf(label -> "suite".equals(label.getName())));
        Allure.feature("REST api (rest-assured)");
        Allure.story("Catalogue");
        Allure.suite("REST api (rest-assured)");

        Map<String, Object> data = new HashMap<String, Object>();
        data.put("year", 2019);
        data.put("price", 1849.99);
        data.put("CPU model", "Intel Core i9");
        data.put("Hard disk size", "1 TB");
        Map<String, Object> payload = new HashMap<String, Object>();
        payload.put("name", "Lenovo notebook");
        payload.put("data", data);

        Response response =
        given().
            filter(new AllureRestAssured()).
            body(payload).
            //contentType("application/json").
            //log().all().
        when().
            post("").
        then().
            log().all().
            statusCode(200).
            body("name", equalTo("Lenovo notebook")).
        extract().response();
        JsonPath body = response.jsonPath();
        String id = (String) body.get("id");
        //try {
        //    Files.writeString(Paths.get("shared.txt"), id);
        //} catch (IOException e) { }
        SharedData.id = id;

        CatalogItem item = response.as(CatalogItem.class);
    }

    @Test
    @Order(2)
    public void get_object() {
        Allure.getLifecycle().updateTestCase(tr -> tr.getLabels().removeIf(label -> "suite".equals(label.getName())));
        Allure.feature("REST api (rest-assured)");
        Allure.story("Catalogue");
        Allure.suite("REST api (rest-assured)");

        String id = null;
        //try {
        //    id = Files.readString(Paths.get("shared.txt"));
        //} catch (IOException e) { }
        id = SharedData.id;
        Response response =
        with().
            filter(new AllureRestAssured()).
        when().
            get("/{id}", id).
        then().
            log().all().
            statusCode(200).
            body("id", equalTo(id)).
            body("name", equalTo("Lenovo notebook")).
        extract().response();

        CatalogItem item = response.as(CatalogItem.class);
    }

    @Test
    @Order(3)
    public void delete_object() {
        Allure.getLifecycle().updateTestCase(tr -> tr.getLabels().removeIf(label -> "suite".equals(label.getName())));
        Allure.feature("REST api (rest-assured)");
        Allure.story("Catalogue");
        Allure.suite("REST api (rest-assured)");

        String id = null;
        //try {
        //    id = Files.readString(Paths.get("shared.txt"));
        //} catch (IOException e) { }
        id = SharedData.id;
        // Delete pet
        with().
            filter(new AllureRestAssured()).
        when().
            delete("/{id}", id).
        then().
            log().all().
            statusCode(200).
            body("message", equalTo("Object with id = " + id + " has been deleted."));

        // Get deleted pet
        with().
            filter(new AllureRestAssured()).
        when().
            get("/{id}", id).
        then().
            log().all().
            statusCode(404).
            body("error", equalTo("Oject with id=" + id + " was not found."));
    }

}

class SharedData {
    public static String id;
}

File: python/pytest/rest_api/tests/catalog_test.py
import pytest
from pytest_bdd import scenarios, given, when, then, parsers
from rest_api.model import catalog


scenarios('../features/catalog.feature')


# Shared variables between scenarios
@pytest.fixture(scope="module")
def context():
    return {
        "id": None,
        "result": None,
        "payload": None
    }


@given('I have a catalog')
def step_impl(context):
    pass


@when('I add an object')
def step_impl(context, report):
    payload = {
        "name": "Lenovo notebook",
        "data": {
            "year": 2019,
            "price": 1849.99,
            "CPU model": "Intel Core i9",
            "Hard disk size": "1 TB"
        }
    }
    report.attach(f"Query: <code>POST</code>", escape_html=False)
    report.attach("Payload:", body=payload, mime=report.Mime.JSON)
    result = catalog.post(payload)
    report.attach(f"Status code: <code>{result.status_code}</code>", escape_html=False)
    report.attach("Response:", body=result.json(), mime=report.Mime.JSON)
    assert result.status_code == 200
    context['id'] = result.json()["id"]
    context['result'] = result
    context['payload'] = payload


@then('The object is added')
def step_impl(context, report):
    id = context['id']
    payload = context['payload']
    report.attach(f"Query: <code>GET /{id}</code>", escape_html=False)
    result = catalog.get(id)
    report.attach(f"Status code: <code>{result.status_code}</code>", escape_html=False)
    report.attach("Response:", body=result.json(), mime=report.Mime.JSON)
    assert result.status_code == 200
    obj = result.json()
    for key in payload:
        assert obj[key] == payload[key]
    context['result'] = result


@given('An object exists in the catalog')
def step_impl(context, report):
    id = context['id']
    catalog.get(id)


@when('I query an object')
def step_impl(context, report):
    id = context['id']
    report.attach(f"Query: <code>GET /{id}</code>", escape_html=False)
    result = catalog.get(id)
    report.attach(f"Status code: <code>{result.status_code}</code>", escape_html=False)
    report.attach("Response:", body=result.json(), mime=report.Mime.JSON)
    context['result'] = result


@then('I get the object information')
def step_impl(context):
    result = context['result']
    assert result.status_code == 200
    assert 'id' in result.json()
    assert 'name' in result.json()
    assert 'data' in result.json()


@when('I delete an object')
def step_impl(context, report):
    id = context['id']
    report.attach(f"Query: <code>DELETE /{id}</code>", escape_html=False)
    result = catalog.delete(id)
    report.attach(f"Status code: <code>{result.status_code}</code>", escape_html=False)
    report.attach("Response:", body=result.json(), mime=report.Mime.JSON)
    assert result.status_code == 200
    assert result.json()['message'] == f"Object with id = {id} has been deleted."
    context['result'] = result


@then('The object is deleted')
def step_impl(context, report):
    id = context['id']
    report.attach(f"Query: <code>GET /{id}</code>", escape_html=False)
    result = catalog.get(id)
    report.attach(f"Status code: <code>{result.status_code}</code>", escape_html=False)
    report.attach("Response:", body=result.json(), mime=report.Mime.JSON)
    assert result.status_code == 404
    assert result.json()['error'] == f"Oject with id={id} was not found."
    context['result'] = result
